src/NumericalAlgorithms/Interpolation/LinearInterpolationND.hpp

```diff
@@ -0,0 +1,374 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <memory>
+
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace intrp {
+
+template <size_t Dimension, size_t NumberOfVariables, bool UniformSpacing>
+class LinearSpanInterpolationND {
+ public:
+  using DataPointer = gsl::span<double const>;
+  //! X values of the table. Only used if allocated
+  std::array<DataPointer, Dimension> x;
+  //! Y values of the table. Only used if allocated
+  DataPointer y;
+
+  std::array<size_t, Dimension> numberOfPoints;
+  std::array<double, Dimension> inverseSpacing, spacing;
+
+  std::array<bool, Dimension> extrapolate_low;
+  std::array<bool, Dimension> extrapolate_high;
+
+ private:
+  //! Determine relative index bracketing for non-uniform spacing
+  inline size_t findIndexGeneral(size_t const whichDimension,
+                                 double const& xin) const {
+    int lowerIndexBound = 0;
+    int upperIndexBound = numberOfPoints[whichDimension] - 1;
+
+    ASSERT((xin > x[whichDimension][lowerIndexBound]) or
+               extrapolate_low[whichDimension],
+           "Interpolation exceeds lower table bounds");
+    ASSERT((xin < x[whichDimension][upperIndexBound]) or
+               extrapolate_high[whichDimension],
+           "Interpolation exceeds upper table bounds");
+
+    while (upperIndexBound - lowerIndexBound > 1) {
+      int currentIndex =
+          lowerIndexBound + (upperIndexBound - lowerIndexBound) / 2;
+      if (xin < x[whichDimension][currentIndex])
+        upperIndexBound = currentIndex;
+      else
+        lowerIndexBound = currentIndex;
+    }
+    return lowerIndexBound;
+  };
+
+  //! Determine relative index bracketing for non-uniform spacing
+  inline size_t findIndexUniform(size_t const whichDimension,
+                                 double const& xin) const {
+    // Compute coordinate relative to lowest table bound
+    const auto relativeCoordinate = (xin - x[whichDimension][0]);
+
+    if (relativeCoordinate == 0) {
+      return 0;
+    }
+
+    // We are exceeding the table bounds:
+    // Use linear extrapolation based of the lowest
+    // two points in the table
+    if (relativeCoordinate < 0) {
+      ASSERT(extrapolate_low[whichDimension],
+             "Interpolation exceeds lower table bounds.");
+    }
+
+    auto currentIndex =
+        static_cast<int>(relativeCoordinate * inverseSpacing[whichDimension]);
+
+    // We are exceeding the table bounds:
+    // Use linear extrapolation based of the highest
+    // two points in the table
+
+    if (currentIndex > numberOfPoints[whichDimension] - 2) {
+      ASSERT(extrapolate_high[whichDimension],
```

Comment:
`ASSERT(relative_coordinate <= x[which_dimension-2], ...);` (maybe I got the index offset wrong here), then move up to below other assert on lower bound